components/swim/src/server/mod.rs

```diff
@@ -0,0 +1,280 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! The SWIM server.
+//!
+//! Creates `Server` structs, that hold everything we need to run the SWIM protocol. Winds up with
+//! 3 separate threads - inbound (incoming connections), outbound (the Probe protocl), and expire
+//! (turning Suspect members into Confirmed members).
+
+pub mod expire;
+pub mod inbound;
+pub mod outbound;
+
+use std::collections::HashSet;
+use std::clone::Clone;
+use std::fmt;
+use std::net::{ToSocketAddrs, UdpSocket, SocketAddr};
+use std::sync::{Arc, RwLock};
+use std::sync::atomic::{AtomicBool, AtomicIsize, Ordering};
+use std::sync::mpsc::channel;
+use std::time::Duration;
+use std::thread;
+
+use error::{Result, Error};
+use member::{Member, Health, MemberList};
+use trace::Trace;
+use rumor::{RumorList, RumorKey};
+
+/// The server struct. Is thread-safe.
+#[derive(Debug)]
+pub struct Server {
+    pub name: Arc<RwLock<String>>,
+    pub socket: UdpSocket,
+    pub member: Arc<RwLock<Member>>,
+    pub member_list: MemberList,
+    pub rumor_list: RumorList,
+    // These are all here for testing support
+    pub pause: Arc<AtomicBool>,
+    pub trace: Arc<RwLock<Trace>>,
+    pub rounds: Arc<AtomicIsize>,
+    pub blacklist: Arc<RwLock<HashSet<SocketAddr>>>,
+}
+
+impl Clone for Server {
+    fn clone(&self) -> Server {
+        let socket = match self.socket.try_clone() {
+            Ok(socket) => socket,
+            Err(e) => {
+                error!("Failed to clone socket; trying again: {:?}", e);
+                return self.clone();
+            }
+        };
+        Server {
+            name: self.name.clone(),
+            socket: socket,
+            member: self.member.clone(),
+            member_list: self.member_list.clone(),
+            rumor_list: self.rumor_list.clone(),
+            pause: self.pause.clone(),
+            trace: self.trace.clone(),
+            rounds: self.rounds.clone(),
+            blacklist: self.blacklist.clone(),
+        }
+    }
+}
+
+impl Server {
+    /// Create a new server, bound to the `addr`, hosting a particular `member`, and with a
+    /// `Trace` struct.
+    ///
+    /// # Errors
+    ///
+    /// * Returns `Error::CannotBind` if the socket cannot be bound
+    /// * Returns `Error::SocketSetReadTimeout` if the socket read timeout cannot be set
+    /// * Returns `Error::SocketSetWriteTimeout` if the socket write timeout cannot be set
+    pub fn new<A: ToSocketAddrs>(addr: A, member: Member, trace: Trace) -> Result<Server> {
+        let socket = match UdpSocket::bind(addr) {
+            Ok(socket) => socket,
+            Err(e) => return Err(Error::CannotBind(e)),
+        };
+        match socket.set_read_timeout(Some(Duration::from_millis(1000))) {
```

Comment:
This would be better written with [map_err](https://doc.rust-lang.org/std/result/enum.Result.html#method.map_err)

```
try!(socket.set_read_timeout(Some(Duration::from_millis(1000))).map_err(|e| Error::SocketSetReadTimeout(e)));
```